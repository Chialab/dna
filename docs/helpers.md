# Helpers

The DNA module exports a set of useful methods too. Sometimes those methods are based on Custom Elements specification, others they are internally used helpers with a completly custom logic.

## get

Like the [`CustomElementRegistry.get`](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/get) method, it returns the component constructor defined for a given tag:

```ts
import { Component, define, get } from '@chialab/dna';

class CardElement extends Component {}

define('x-card', Card);

get('x-card') // -> Card
```

## whenDefined

Like the [`CustomElementRegistry.whenDefined`](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/whenDefined) method, it returns a Promise that resolves once a component has been defined with the name:

```ts
import { whenDefined } from '@chialab/dna';

whenDefined('x-card')
    .then(() => {
        // -> The x-card has been defined
    });

import('./x-card.js');
```

## upgrade

Like the [`CustomElementRegistry.upgrade`](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/upgrade) method, it is used to upgrade nodes that had been created before element definition:

```ts
import { Component, DOM, define, upgrade } from '@chialab/dna';

class Card extends Component {}

const element = DOM.createElement('x-card');
console.log(Object.getPrototypeOf(element)); // -> HTMLElement

define('x-card', Card);
upgrade(element);

console.log(Object.getPrototypeOf(element)); // -> Card
```

All children and descendants of the passed element will be upgraded too, so `upgrade` can be used for tag re-hydration after a server side rendering:

```ts
import { upgrade } from '@chialab/dna';

document.addEventListener('DOMContentLoaded', (event) => {
    upgrade(document.body);
});
```

## extend

This function extends a native `HTMLElement` constructor with all DNA prototype capabilities. This is useful to extend native constructors instead of using the default `Component` class:

```ts
import { extend, DOM } from '@chialab/dna';

// mix DNA features with the base class
export class AnchorComponent extends extend(DOM.window.HTMLAnchorElement) {
    ...
}

// the component inherit the HTMLAnchorElement's href behavior
const a = new AnchorComponent();
a.setAttribute('href', '/demo');

console.log(a.href); // -> https://my.domain/demo
```

<aside class="note">

Use the `DOM.window` namespace instead of the global constructor to make sure to write Node compatible components.

</aside>

## css

The `css` helper is a method internally used by DNA convert a CSS string into its scoped Shadow DOM version. This can be used to add extra manipulation to the CSS string:

```ts
import { DOM, Component, html, css, define } from '@chialab/dna';

const cssText = css('x-article', 'h1 { color: red; }')
        .replace(/red/g, 'blue');

const style = DOM.document.createElement('style');
style.textContent = cssText;

DOM.appendChild(DOM.document.head, style);
```

## interpolate

The `interpolate` helper compiles a string expression like `Hello {{ nameÂ }}` into static values. It is internally used to provide interpolation to a template generated from `<template>` element.

```ts
import { interpolate } from '@chialab/dna';

interpolate(`Hello, my name is {{firstName.toUpperCase()}} and I'm {{age * 2}} years old. I was born on {{birthdate.toLocaleDateString()}}`, {
    name: 'Alan',
    age: 20,
    birthdate: new Date('05/05/1993'),
});
// -> Hello, my name is Alan and I'm 40 years old. I was born on 05/05/1993.
```

<aside class="note">

Since the `interpolate` method generates a function which evaluate any script inside the interpolation symbols, you must make sure it is safe code. Furthemore, interpolated scripts will not be transpiled by Babel or any other transpiler, so any unsupported syntax by the browser will throw a `SyntaxError`.

</aside>

## Creating an adapter

An adapter is a compatibility layer between DNA and the environment where it runs. Since DNA is completely based on the DOM, in order to create an adapter we need to replace the methods of the `DOM` namespace:

```ts
import { DOM } from '@chialab/dna';

Object.assign(DOM, {
    window: windowImplementation,
    document: documentImplementation,
    createElement(tagName: string, options?: ElementCreationOptions): Element { ... },
    createElementNS(namespaceURI: string, tagName: string): Element { ... },
    createTextNode(data: string): Text { ... },
    createEvent(typeArg: string, eventInitDict?: CustomEventInit): CustomEvent { ... },
    appendChild(parent: Element, newChild: Node): Node { ... },
    removeChild(parent: Element, oldChild: Node): Node { ... },
    insertBefore(parent: Element, newChild: Node, refChild: Node | null): Node { ... },
    replaceChild(parent: Element, newChild: Node, oldChild: Node): Node { ... },
    getAttribute(element: Element, qualifiedName: string): string | null { ... },
    hasAttribute(element: Element, qualifiedName: string): boolean { ... },
    setAttribute(element: Element, qualifiedName: string, value: string): void { ... },
    removeAttribute(element: Element, qualifiedName: string) { ... },
    matches(element: Element, selectorString: string): boolean { ... },
    dispatchEvent(element: Node, event: Event): boolean { ... },
});
```
