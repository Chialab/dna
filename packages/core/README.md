[![Logo](https://cdn.rawgit.com/Chialab/dna/master/logo.svg)](http://chialab.io/p/dna)

Evolution-based components.

[Documentation](http://chialab.io/p/dna/docs) | [Issue tracker](https://github.com/Chialab/dna/issues) | [Project home page](http://chialab.io/p/dna) | [Author home page](https://www.chialab.it)

## Install

[![NPM](https://img.shields.io/npm/v/@dnajs/core.svg)](https://www.npmjs.com/package/@dnajs/core)
```
$ npm i @dnajs/core --save
```

## Usage

TODO
