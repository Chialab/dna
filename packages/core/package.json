{
  "name": "@dnajs/core",
  "version": "2.7.6",
  "description": "Evolution-based components.",
  "author": "Chialab <dev@chialab.io> (http://www.chialab.it)",
  "main": "dist/core.js",
  "module": "index.js",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/chialab/dna.git"
  },
  "keywords": [
    "DNA",
    "components",
    "Chialab",
    "webcomponents"
  ],
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/chialab/dna/issues"
  },
  "homepage": "https://github.com/chialab/dna#readme"
}
