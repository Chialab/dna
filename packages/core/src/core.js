export { default as DOM } from './lib/dom.js';
export { mix } from './lib/mixins.js';
export { prop } from './lib/property.js';
export { render } from './lib/render.js';
