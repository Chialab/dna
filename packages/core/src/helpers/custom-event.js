let CustomEvent = self.CustomEvent;

try {
    // eslint-disable-next-line
    new CustomEvent('test');
} catch(ex) {
    CustomEvent = function(ev, params) {
        let evt = document.createEvent('CustomEvent');
        evt.initCustomEvent(ev, params.bubbles, params.cancelable, params.detail);
        return evt;
    };
    CustomEvent.prototype = self.CustomEvent.prototype;
}

/**
 * Polyfill `CustomEvent`.
 * @type {Function}
 * @private
 */
export { CustomEvent };
