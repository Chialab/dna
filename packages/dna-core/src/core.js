import * as DOM_HELPERS from './lib/dom.js';
import { ComponentMixin } from './mixins/component.js';
import { PropertiesMixin } from './mixins/properties-component.js';
import { EventsMixin } from './mixins/events-component.js';
import { StyleMixin } from './mixins/style-component.js';
import { TemplateMixin } from './mixins/template-component.js';

/**
 * A set of DOM helpers for callbacks trigger when Custom Elements
 * are not supported by the browser.
 * @name DOM
 * @namespace DOM
 * @memberof! DNA.
 * @static
 */
export const DOM = DOM_HELPERS;
/**
 * A set of core mixins.
 * @name MIXINS
 * @namespace MIXINS
 * @memberof! DNA.
 * @static
 */
export const MIXINS = {
    ComponentMixin,
    PropertiesMixin,
    EventsMixin,
    StyleMixin,
    TemplateMixin,
};
export { mix } from './lib/mixins.js';
export { prop } from './lib/property.js';
export * from './lib/symbols.js';
