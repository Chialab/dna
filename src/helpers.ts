import { document, CustomEvent, Node, HTMLElement, Event } from './window';
const { DOCUMENT_NODE, TEXT_NODE, ELEMENT_NODE } = Node;


/**
 * Alias to Array.isArray.
 */
export const isArray = Array.isArray;

/**
 * Alias to Object.getOwnPropertyDescriptor.
 */
export const getOwnPropertyDescriptor = Object.getOwnPropertyDescriptor;

/**
 * Alias to Object.defineProperty.
 */
export const defineProperty = Object.defineProperty;

/**
 * Alias to Node.prototype.appendChild.
 */
export const appendChildImpl = Node.prototype.appendChild;

/**
 * Alias to Node.prototype.removeChild.
 */
export const removeChildImpl = Node.prototype.removeChild;

/**
 * Alias to Node.prototype.insertBefore.
 */
export const insertBeforeImpl = Node.prototype.insertBefore;

/**
 * Alias to Node.prototype.replaceChild.
 */
export const replaceChildImpl = Node.prototype.replaceChild;

/**
 * Alias to Node.prototype.isConnected.
 */
export const isConnectedImpl = getOwnPropertyDescriptor(Node.prototype, 'isConnected');

/**
 * Alias to HTMLElement.prototype.getAttribute.
 */
export const getAttributeImpl = HTMLElement.prototype.getAttribute;

/**
 * Alias to HTMLElement.prototype.hasAttribute.
 */
export const hasAttributeImpl = HTMLElement.prototype.hasAttribute;

/**
 * Alias to HTMLElement.prototype.setAttribute.
 */
export const setAttributeImpl = HTMLElement.prototype.setAttribute;

/**
 * Alias to HTMLElement.prototype.matches.
 */
export const matchesImpl = HTMLElement.prototype.matches || HTMLElement.prototype.webkitMatchesSelector || (HTMLElement.prototype as any).msMatchesSelector as typeof Element.prototype.matches;

/**
 * Create a Custom Event.
 * @param typeArg The name of the event.
 * @param eventInitDict The options of the event.
 * @return The constructed Custom Event.
 */
export const createEventImpl = (typeArg: string, eventInitDict: CustomEventInit<unknown> = {}): CustomEvent<unknown> => {
    let event;
    try {
        event = new CustomEvent(typeArg, eventInitDict);
    } catch {
        event = document.createEvent('CustomEvent');
        event.initCustomEvent(typeArg, eventInitDict.bubbles || false, eventInitDict.cancelable || false, eventInitDict.detail);
    }
    return event;
};

/**
 * Check if a node is a Document instance.
 * @param node The node to check.
 * @return The node is a Document instance.
 */
export const isDocument = (node: any): node is Document => node && node.nodeType === DOCUMENT_NODE;

/**
 * Check if a node is a Text instance.
 * @param node The node to check.
 * @return The node is a Text instance.
 */
export const isText = (node: any): node is Text => node && node.nodeType === TEXT_NODE;

/**
 * Check if a node is an Element instance.
 * @param node The node to check.
 * @return The node is an Element instance.
 */
export const isElement = (node: any): node is HTMLElement => node && node.nodeType === ELEMENT_NODE;

/**
 * Check if an object is an Event instance.
 * @param node The node to check.
 * @return The object is an Event instance.
 */
export const isEvent = (event: any): event is Event => event instanceof Event;

