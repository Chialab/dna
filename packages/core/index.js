/**
 * DNA
 * (c) 2015-2017 Chialab (http://www.chialab.it) <dev@chialab.io>
 * http://dna.chialab.io
 *
 * Evolution-based components.
 */
import { mix, MIXINS } from './src/core.js';
import { registry } from './src/lib/registry.js';
import { proxy } from './src/lib/proxy.js';

/**
 * @namespace DNA
 */
export * from './src/core.js';
export { registry };
export { proxy };
export { bootstrap } from './src/lib/bootstrap.js';
export { define } from './src/lib/define.js';
export { render } from './src/lib/render.js';

const Component = proxy(class {});

/**
 * Simple Custom Component with some behaviors.
 * @class BaseComponent
 * @memberof DNA.
 * @static
 *
 * @example
 * ```js
 * // my-component.js
 * import { BaseComponent } from '@dnajs/core';
 * export class MyComponent extends BaseComponent {
 *   static get observedAttributes() {
 *     return ['...', '...'];
 *   }
 *   get css() {
 *     return '...';
 *   }
 *   get events() {
 *     return {
 *       '...': '...'
 *     };
 *   }
 *   get template() {
 *     return '...';
 *   }
 *   get properties() {
 *     return { ... };
 *   }
 * }
 * ```
 */
export class BaseComponent extends mix(Component).with(
    MIXINS.ComponentMixin,
    MIXINS.PropertiesMixin,
    MIXINS.StyleMixin,
    MIXINS.EventsMixin,
    MIXINS.TemplateMixin
) {
    constructor(node) {
        super();
        if (!node) {
            let desc = registry.get(this.is, true);
            let config = desc.config;
            node = document.createElement(
                config.extends ? config.extends : desc.is
            );
            if (config.extends) {
                node.setAttribute('is', desc.is);
            }
        }
        this.node = node;
    }
}
