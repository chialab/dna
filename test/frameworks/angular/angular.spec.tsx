import 'zone.js';
import 'zone.js/testing';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { ÉµgetCleanupHook as getCleanupHook, getTestBed } from '@angular/core/testing';
import { BrowserTestingModule, platformBrowserTesting } from '@angular/platform-browser/testing';
import { render } from '@testing-library/angular';
import { afterEach, beforeAll, beforeEach, describe, expect, test, vi } from 'vitest';
import { IS_BROWSER } from '../../helpers';
import '../../polyfills';
import {
    defineTestElements,
    type TestElement1,
    type TestElement2,
    type TestElement3,
    type TestElement4,
    type TestElement5,
    type TestElement6,
} from '../TestElements';
import '../TestElements';

describe.runIf(IS_BROWSER)('Angular', () => {
    beforeAll(() => {
        getTestBed().resetTestEnvironment();
        getTestBed().initTestEnvironment(BrowserTestingModule, platformBrowserTesting());
    });
    beforeEach(getCleanupHook(false));
    afterEach(getCleanupHook(true));

    test('should update text content', async () => {
        const { container, rerender } = await render('<test-frameworks-1>{{ text }}</test-frameworks-1>', {
            imports: [BrowserTestingModule],
            schemas: [CUSTOM_ELEMENTS_SCHEMA],
            componentProperties: {
                text: 'Text',
            },
        });

        const element = container.children[0] as TestElement1;
        expect(element.slotChildNodes).toHaveLength(1);
        expect(element.childNodesBySlot(null)).toHaveLength(1);
        expect(element.childNodesBySlot('children')).toHaveLength(0);
        expect(element.childNodes[0]).toHaveProperty('textContent', 'Text');
        expect(container.innerHTML).toBe(
            '<test-frameworks-1 :scope="test-frameworks-1" :defined=""><span>Text</span><div></div></test-frameworks-1>'
        );

        await rerender({
            componentProperties: { text: 'Update' },
        });

        expect(element.slotChildNodes).toHaveLength(1);
        expect(element.childNodesBySlot(null)).toHaveLength(1);
        expect(element.childNodesBySlot('children')).toHaveLength(0);
        expect(element.childNodes[0]).toHaveProperty('textContent', 'Update');
        expect(container.innerHTML).toBe(
            '<test-frameworks-1 :scope="test-frameworks-1" :defined=""><span>Update</span><div></div></test-frameworks-1>'
        );
    });

    test('should update text content with multiple text nodes', async () => {
        const { container, rerender } = await render('<test-frameworks-1>{{ text }} children</test-frameworks-1>', {
            schemas: [CUSTOM_ELEMENTS_SCHEMA],
            componentProperties: {
                text: 'Text',
            },
        });

        const element = container.children[0] as TestElement1;
        expect(element.slotChildNodes).toHaveLength(1);
        expect(element.childNodesBySlot(null)).toHaveLength(1);
        expect(element.childNodesBySlot('children')).toHaveLength(0);
        expect(element.childNodes[0]).toHaveProperty('textContent', 'Text children');
        expect(element.childNodes[0].childNodes[0]).toHaveProperty('textContent', 'Text children');
        expect(container.innerHTML).toBe(
            '<test-frameworks-1 :scope="test-frameworks-1" :defined=""><span>Text children</span><div></div></test-frameworks-1>'
        );

        await rerender({
            componentProperties: { text: 'Update' },
        });

        expect(element.slotChildNodes).toHaveLength(1);
        expect(element.childNodesBySlot(null)).toHaveLength(1);
        expect(element.childNodesBySlot('children')).toHaveLength(0);
        expect(element.childNodes[0]).toHaveProperty('textContent', 'Update children');
        expect(element.childNodes[0].childNodes[0]).toHaveProperty('textContent', 'Update children');
        expect(container.innerHTML.replace(/\n\s*/g, ' ')).toBe(
            '<test-frameworks-1 :scope="test-frameworks-1" :defined=""><span>Update children</span><div></div></test-frameworks-1>'
        );
    });

    test('should update named slots', async () => {
        const { container, rerender } = await render(
            `<test-frameworks-1>
    Text
    <h1 *ngIf="title" slot="children">Title</h1>
    <h2 *ngIf="!title" slot="children">Subtitle</h2>
    end
</test-frameworks-1>`,
            {
                schemas: [CUSTOM_ELEMENTS_SCHEMA],
                componentProperties: {
                    title: true,
                },
            }
        );

        const element = container.children[0] as TestElement1;
        const textNode = element.childNodes[0].childNodes[0];
        expect(element.childNodesBySlot('children')).toHaveLength(1);
        expect(element.childNodesBySlot('children')[0]).toHaveProperty('tagName', 'H1');
        expect(container.innerHTML.replace(/\n\s*/g, ' ')).toBe(
            '<test-frameworks-1 :scope="test-frameworks-1" :defined=""><span> Text  end </span><div><h1 slot="children">Title</h1></div></test-frameworks-1>'
        );

        await rerender({
            componentProperties: { title: false },
        });

        expect(element.childNodesBySlot('children')).toHaveLength(1);
        expect(element.childNodesBySlot('children')[0]).toHaveProperty('tagName', 'H2');
        expect(element.childNodes[0].childNodes[0]).toBe(textNode);
        expect(container.innerHTML.replace(/\n\s*/g, ' ')).toBe(
            '<test-frameworks-1 :scope="test-frameworks-1" :defined=""><span> Text  end </span><div><h2 slot="children">Subtitle</h2></div></test-frameworks-1>'
        );
    });

    test('mixed slots', async () => {
        const { container, rerender } = await render(
            `<test-frameworks-1>
    <span>Test</span>
    <h1 *ngIf="title" slot="children">Title</h1>
    <span>Test</span>
    <h2 *ngIf="title" slot="children">Title</h2>
    <span>Test</span>
</test-frameworks-1>`,
            {
                schemas: [CUSTOM_ELEMENTS_SCHEMA],
                componentProperties: {
                    title: false,
                },
            }
        );

        const element = container.children[0] as TestElement1;
        expect(element.slotChildNodes).toHaveLength(5);
        expect(element.childNodesBySlot(null)).toHaveLength(3);
        expect(element.childNodesBySlot('children')).toHaveLength(0);

        await rerender({ componentProperties: { title: true } });

        expect(element.slotChildNodes).toHaveLength(7);
        expect(element.childNodesBySlot(null)).toHaveLength(3);
        expect(element.childNodesBySlot('children')).toHaveLength(2);
        expect(element.childNodes[0].childNodes[0]).toHaveProperty('textContent', 'Test');
        expect(element.childNodes[0].childNodes[1]).toHaveProperty('textContent', 'Test');
        expect(element.childNodes[0].childNodes[2]).toHaveProperty('textContent', 'Test');
        expect(element.childNodes[1].childNodes[0]).toHaveProperty('tagName', 'H1');
        expect(element.childNodes[1].childNodes[0]).toHaveProperty('textContent', 'Title');
        expect(element.childNodes[1].childNodes[1]).toHaveProperty('tagName', 'H2');
        expect(element.childNodes[1].childNodes[1]).toHaveProperty('textContent', 'Title');

        await rerender({ componentProperties: { title: false } });

        expect(element.slotChildNodes).toHaveLength(5);
        expect(element.childNodesBySlot(null)).toHaveLength(3);
        expect(element.childNodesBySlot('children')).toHaveLength(0);
    });

    test('nested slot', async () => {
        const { container, rerender } = await render(
            `<test-frameworks-2>
    <h1 *ngIf="title" slot="title">Title</h1>
    <img src="data:image/png;base64," alt="" />
    <p>Body</p>
</test-frameworks-2>`,
            {
                schemas: [CUSTOM_ELEMENTS_SCHEMA],
                componentProperties: {
                    title: false,
                },
            }
        );

        const element = container.children[0] as TestElement2;
        expect(element.childNodesBySlot(null)).toHaveLength(2);
        expect(element.childNodesBySlot('title')).toHaveLength(0);
        expect(element.children).toHaveLength(2);
        expect(element.children[0]).toHaveProperty('tagName', 'DIV');
        expect(element.children[0]).toHaveProperty('className', 'layout-header');
        expect(element.children[0].children).toHaveLength(1);
        expect(element.children[0].children[0]).toHaveProperty('tagName', 'test-frameworks-title'.toUpperCase());
        expect(element.children[0].children[0].children[0]).toHaveProperty('tagName', 'SPAN');
        expect(element.children[0].children[0].children[0]).toHaveProperty('textContent', 'Untitled');

        await rerender({ componentProperties: { title: true } });

        expect(element.childNodesBySlot(null)).toHaveLength(2);
        expect(element.childNodesBySlot('title')).toHaveLength(1);
        expect(element.childNodesBySlot('title')[0]).toHaveProperty('tagName', 'H1');
        expect(element.children[0].children[0].children[0]).toHaveProperty('tagName', 'SPAN');
        expect(element.children[0].children[0].children[0]).toHaveProperty('textContent', 'Title');
        expect(element.children[1]).toHaveProperty('tagName', 'DIV');
        expect(element.children[1]).toHaveProperty('className', 'layout-body');
        expect(element.children[1].children[0]).toHaveProperty('tagName', 'IMG');
        expect(element.children[1].children[0]).toHaveProperty('src', 'data:image/png;base64,');
        expect(element.children[1].children[1]).toHaveProperty('tagName', 'P');
        expect(element.children[1].children[1]).toHaveProperty('textContent', 'Body');
    });

    test('slot moved across elements', async () => {
        const { container, rerender } = await render(
            `<test-frameworks-3 [collapsed]="collapsed">
    <h1>Title</h1>
    <img src="data:image/png;base64," alt="" />
    <p>Body</p>
</test-frameworks-3>`,
            {
                schemas: [CUSTOM_ELEMENTS_SCHEMA],
                componentProperties: {
                    collapsed: false,
                },
            }
        );
        defineTestElements();

        const element = container.children[0] as TestElement3;
        window.customElements.upgrade(element);
        expect(element.children).toHaveLength(1);
        expect(element.children[0].children[0]).toHaveProperty('tagName', 'H1');
        expect(element.children[0].children[0]).toHaveProperty('textContent', 'Title');
        expect(element.children[0].children[1]).toHaveProperty('tagName', 'IMG');
        expect(element.children[0].children[1]).toHaveProperty('src', 'data:image/png;base64,');
        expect(element.children[0].children[2]).toHaveProperty('tagName', 'P');
        expect(element.children[0].children[2]).toHaveProperty('textContent', 'Body');

        await rerender({ componentProperties: { collapsed: true } });

        expect(element.children[0]).toHaveProperty('tagName', 'H1');
        expect(element.children[0]).toHaveProperty('textContent', 'Title');
        expect(element.children[1]).toHaveProperty('tagName', 'IMG');
        expect(element.children[1]).toHaveProperty('src', 'data:image/png;base64,');
        expect(element.children[2]).toHaveProperty('tagName', 'P');
        expect(element.children[2]).toHaveProperty('textContent', 'Body');

        await rerender({ componentProperties: { collapsed: false } });

        expect(element.children[0].children[0]).toHaveProperty('tagName', 'H1');
        expect(element.children[0].children[0]).toHaveProperty('textContent', 'Title');
        expect(element.children[0].children[1]).toHaveProperty('tagName', 'IMG');
        expect(element.children[0].children[1]).toHaveProperty('src', 'data:image/png;base64,');
        expect(element.children[0].children[2]).toHaveProperty('tagName', 'P');
        expect(element.children[0].children[2]).toHaveProperty('textContent', 'Body');
    });

    test('slot moved and replaced', async () => {
        const { container, rerender } = await render(
            `<test-frameworks-4 [switch]="switchValue">
    <ng-container *ngIf="switchValue">World</ng-container>
    <ng-container *ngIf="!switchValue">Hello</ng-container>
</test-frameworks-4>`,
            {
                schemas: [CUSTOM_ELEMENTS_SCHEMA],
                componentProperties: {
                    switchValue: false,
                },
            }
        );

        const element = container.children[0] as TestElement4;
        expect(element.querySelector('.parent-1')).toHaveProperty('textContent', 'Empty');
        expect(element.querySelector('.parent-2')).toHaveProperty('textContent', 'Hello');

        await rerender({ componentProperties: { switchValue: true } });

        expect(element.querySelector('.parent-1')).toHaveProperty('textContent', 'World');
        expect(element.querySelector('.parent-2')).toHaveProperty('textContent', 'Empty');
    });

    test('autonomous properties', async () => {
        const onClick = vi.fn((event) => event.preventDefault());
        const onStringChange = vi.fn();
        const { rerender, container } = await render(
            `<test-frameworks-5
    [stringProp]="stringProp"
    [booleanProp]="booleanProp"
    [numericProp]="numericProp"
    [objectProp]="objectProp"
    (click)="onClick?.($event)"
    (stringchange)="onStringChange?.($event)"
    data-attr="test"
></test-frameworks-5>`,
            {
                componentProperties: {
                    'stringProp': 'test',
                    'booleanProp': true,
                    'numericProp': 1,
                    'objectProp': { test: true },
                },
            }
        );

        const element = container.children[0] as TestElement5;
        expect(element.stringProp).toBe('test');
        expect(element.booleanProp).toBe(true);
        expect(element.numericProp).toBe(1);
        expect(element.objectProp).toEqual({ test: true });
        expect(element.defaultValue).toBe(0);
        expect(element.getAttribute('data-attr')).toBe('test');
        expect(onStringChange).not.toHaveBeenCalled();
        expect(onClick).not.toHaveBeenCalled();
        await rerender({
            componentProperties: {
                onClick,
                onStringChange,
                stringProp: 'changed',
            },
        });
        expect(element.stringProp).toBe('changed');
        expect(onStringChange).toHaveBeenCalledOnce();
        element.click();
        expect(onClick).toHaveBeenCalledOnce();
    });

    /** @TODO https://github.com/angular/angular/issues/63174 */
    test.skip('builtin properties', async () => {
        const onClick = vi.fn((event) => event.preventDefault());
        const onStringChange = vi.fn();
        const { rerender, container } = await render(
            `<a
    is="test-frameworks-6"
    [stringProp]="stringProp"
    [booleanProp]="booleanProp"
    [numericProp]="numericProp"
    [objectProp]="objectProp"
    (click)="onClick?.($event)"
    (stringchange)="onStringChange?.($event)"
    data-attr="test"
></a>`,
            {
                componentProperties: {
                    'stringProp': 'test',
                    'booleanProp': true,
                    'numericProp': 1,
                    'objectProp': { test: true },
                },
            }
        );

        const element = container.children[0] as TestElement6;
        expect(element.stringProp).toBe('test');
        expect(element.booleanProp).toBe(true);
        expect(element.numericProp).toBe(1);
        expect(element.objectProp).toEqual({ test: true });
        expect(element.defaultValue).toBe(0);
        expect(element.getAttribute('data-attr')).toBe('test');
        expect(onStringChange).not.toHaveBeenCalled();
        expect(onClick).not.toHaveBeenCalled();
        await rerender({
            componentProperties: {
                onClick,
                onStringChange,
                stringProp: 'changed',
            },
        });
        expect(element.stringProp).toBe('changed');
        expect(onStringChange).toHaveBeenCalledOnce();
        element.click();
        expect(onClick).toHaveBeenCalledOnce();
    });
});
