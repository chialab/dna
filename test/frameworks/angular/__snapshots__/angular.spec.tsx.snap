// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Angular > should update named slots 1`] = `
"<test-frameworks-1 :scope="test-frameworks-1" :defined=""><span> Text  end
</span><div><h1 slot="children">Title</h1></div></test-frameworks-1>"
`;

exports[`Angular > should update named slots 2`] = `
"<test-frameworks-1 :scope="test-frameworks-1" :defined=""><span> Text  end
</span><div><h2 slot="children">Subtitle</h2></div></test-frameworks-1>"
`;

exports[`Angular > should update text content 1`] = `"<test-frameworks-1 :scope="test-frameworks-1" :defined=""><span>Text</span><div></div></test-frameworks-1>"`;

exports[`Angular > should update text content 2`] = `"<test-frameworks-1 :scope="test-frameworks-1" :defined=""><span>Update</span><div></div></test-frameworks-1>"`;

exports[`Angular > should update text content with multiple text nodes 1`] = `"<test-frameworks-1 :scope="test-frameworks-1" :defined=""><span>Text children</span><div></div></test-frameworks-1>"`;

exports[`Angular > should update text content with multiple text nodes 2`] = `"<test-frameworks-1 :scope="test-frameworks-1" :defined=""><span>Update children</span><div></div></test-frameworks-1>"`;
